druid/src/widget/style_child.rs

```diff
@@ -0,0 +1,78 @@
+// Copyright 2018 The xi-editor Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A widget that themes its child.
+
+use crate::{
+    BaseState, BoxConstraints, Data, Env, Event, EventCtx, LayoutCtx, PaintCtx, Rect, Size,
+    UpdateCtx, Widget, WidgetPod,
+};
+
+use crate::kurbo::Point;
+
+/// A widget that accepts a closure to update the theme for its child.
+pub struct StyleChild<T: Data, F: Fn(&mut Env) + 'static> {
+    f: F,
+    child: WidgetPod<T, Box<dyn Widget<T>>>,
+}
+
+impl<T: Data, F: Fn(&mut Env) + 'static> StyleChild<T, F> {
+    /// Create a widget that themes its child.
+    ///
+    /// Accepts a closure that sets theme values:
+    /// ```
+    /// |env| {
+    ///     env.set(theme::LABEL_COLOR, Color::WHITE);
+    /// }
+    /// ```
+    pub fn new(f: F, child: impl Widget<T> + 'static) -> StyleChild<T, F> {
+        StyleChild {
+            f,
+            child: WidgetPod::new(child).boxed(),
+        }
+    }
+}
+
+impl<T: Data, F: Fn(&mut Env) + 'static> Widget<T> for StyleChild<T, F> {
+    fn paint(&mut self, paint_ctx: &mut PaintCtx, _base_state: &BaseState, data: &T, env: &Env) {
+        let mut new_env = env.clone();
```

Comment:
This is absolutely one of the missing foundational widgets, so thanks for making it! It does need to apply the env consistently across all widget trait methods though, not just paint.

Env might influence behavior - so needs to be in `event`. The `update` method might build text layouts which might depend on style, so needs to be there. The `layout` method might depend on dimensions which are available in the style.

Reply:
Okay! Duplicating this across all the methods won't duplicate the work?

Reply:
Yes, it will duplicate the work. I have some ideas for how to make updates to the env data structure more efficient, but I think we first want to make sure we have the semantics right, then figure out how much of a performance problem we've got, and optimize based on priorities from empirical measuring.